### ISSUE TITLE
Treatment of Systematics Effects in 3x2 point statistics

### ISSUE THREAD

**Issue by chrgeorgiou** [LSSTDESC/firecrown#174]
Currently, the only way to treat systematic effects in 3x2 point statistics is through modifying the variables of a ```ccl.Tracer``` object. For systematic effects that require the calculation of a specific power spectrum that is different from the standard matter power spectrum (e.g. extended intrinsic alignment models such as TATT and Halo model, treatment of baryons, non-linear galaxy bias but also maybe extended cosmologies) the current architecture does not allow their treatment. Below is a code that demonstrates what this would look like with the TATT intrinsic alignment model using CCL. 

```
import numpy as np
import pyccl as ccl
import pyccl.nl_pt as pt
cosmo = ccl.Cosmology(Omega_c=0.27, Omega_b=0.045, h=0.67, A_s=2.1e-9, n_s=0.96)
ell = np.unique(np.geomspace(2,1000,100).astype(int)).astype(float)
# redshift distribution
z = np.linspace(0, 1.5, 1024)
nz = np.exp(-((z-0.7)/0.1)**2)

# Systematics effect is controlled with parameters c_1, c_2, c_d.
# Biases for IAs. Will be converted to the input c_IA values below.
a_1 = 1.
a_2 = 0.5
a_d = 0.5
c_1, c_d, c_2 = pt.translate_IA_norm(cosmo, z, a1=a_1, a1delta=a_d, a2=a_2, Om_m2_for_c2 = False)

# Code that creates a Pk2D object:
ptc = pt.PTCalculator(with_NC=True, with_IA=True,
                      log10k_min=-4, log10k_max=2, nk_per_decade=20)
ptt_i = pt.PTIntrinsicAlignmentTracer(c1=(z,c_1), c2=(z,c_2), cdelta=(z,c_d))
ptt_m = pt.PTMatterTracer()
# IAs x matter
pk_im = pt.get_pt_pk2d(cosmo, ptt_i, tracer2=ptt_m, ptc=ptc)

# Code that computes effect from IA using that Pk2D object
t_lens = ccl.WeakLensingTracer(cosmo, dndz=(z, nz))
t_ia = ccl.WeakLensingTracer(cosmo, dndz=(z, nz), has_shear=False, ia_bias=(z, np.ones_like(z)), use_A_ia=False)
cl_GI = ccl.angular_cl(cosmo, t_lens, t_ia, ell, p_of_k_a=pk_im)
# The weak gravitational lensing power spectrum
cl_GG = ccl.angular_cl(cosmo, t_lens, t_lens, ell)
# The observed angular power spectrum is the sum of the two.
cl_theory = cl_GG+cl_GI # normally we would also have a third term, +cl_II).
```

We see that several projects require the systematics to be incorporated into a ```ccl.Pk2D``` object that will be parsed to ```ccl.angular_cl``` function, instead of having it be in the ```ccl.Tracer``` object. A functionality that allows for this would be very useful to many different projects.

**Comment by tilmantroester** [comment]
Here is a proposal to support likelihood-level systematics, such as systematics that change or add 3D power spectra.

This is with view on two-point analyses but I think the framework should be general enough to support other analyses as well. The basic idea is to populate the `Cosmology._pk_nl` dict with the relevant quantities, such as intrinsic-matter, or galaxy-matter power spectra, such that they can be passed as the `p_of_k_a` argument to `pyccl.angular_cl`.

During `Likelihood.compute_loglike`, apply the likelihood systematics, populating a `CosmologyArgs` container, similar to what is being done in `Source.create_tracer`. The `CosmologyArgs` would store a dict with `Pk2D` objects.
Then create a new `Cosmology` object, based on the input `Cosmology` and the `CosmologyArgs` object.

Finally, pass the `p_of_k_a` argument to `pyccl.angular_cl`, choosing the appropriate power spectrum based on the statistic.

This would support models like TATT (by creating "intrinsic-matter" and "intrinsic-intrinsic" power spectra) and baryons (by adding or replacing the "delta_matter:delta_matter" power spectrum). I think it should also work for perturbative galaxy bias models that are built on top of the `pyccl.nl_pt` infrastructure but this requires some feedback to confirm.

A sketch of a possible API and implementation is below:

Basic usage:
```
tatt_ia = TATTIntrinsicAlignmentModel()
likelihood = ConstGaussian(systematics=[tatt_ia])
```

Support likelihood-level systematics:
```
@dataclass(frozen=True)
class CosmologyArgs:
    """Class for cosmology builder argument."""

    # ...
    p_of_k_a: Dict[str, Pk2D]

class Likelihood:
    def __init__(self, systematics=None):
        # ...

    def compute_loglike(self, cosmo):
        # ...

        for systematic in self.systematics:
            cosmo_args = systematic.apply(cosmo, cosmo_args)
            
        cosmo = pyccl.CosmologyCalculator(cosmo, cosmo_args)

        # ...
```

Specify which power spectrum to use.
```
def _cached_angular_cl(cosmo, tracers, ells, pofk_name):
    return pyccl.angular_cl(cosmo, *tracers, np.array(ells), p_of_k_a=pofk_name)
```

Define the systematic:
```
class TATTIntrinsicAlignmentModel(Systematic):
    def __init__(self):
        self.pt_calculator = pt.PTCalculator(with_NC=True, with_IA=True,
                      log10k_min=-4, log10k_max=2, nk_per_decade=20)
        self.z = z

    def apply(self, cosmo, cosmo_args):
        # ...
        ptt_i = pt.PTIntrinsicAlignmentTracer(c1=(self.z, c_1), c2=(self.z, c_2), cdelta=(self.z, c_d))
        ptt_m = pt.PTMatterTracer()
        pk_im = pt.get_pt_pk2d(cosmo, ptt_i, tracer2=ptt_m, ptc=self.pt_calculator)

        cosmo_args.p_of_k_a["i:delta_matter"] = pk_im

        return cosmo_args
```

**Comment by marcpaterno** [comment]
Tilman, can you describe a bit more in words what you mean for "systematics" to represent? For example,  do they have a place in all types of likelihoods, or only Gaussian-related likelihoods (ones related to a difference between a prediction vector and a theory vector, rather than, say, a Poisson likelihood), or something else?

**Comment by jablazek** [comment]
@damonge : can you clarify what you mean by "pass the Pk2ds directly rather than through a name"?

@tilmantroester : I believe that PT models for nonlinear galaxy bias should work with the same method you sketch out above. There are some technical points about how many different Pk2ds we want to store. For instance, we could throw all of the components for PT models (for IA, bias, and the cross terms) into the dictionary. Then varying the different bias/IA parameters is free, since they just multiply the Pk terms. On the other hand, of only the II, GG, and GI terms are stored, you have to recalculate when changing the bias/IA params. There may be a slick way to store these quantities under the hood to allow fast sampling.

One conceptual point, perhaps related to @marcpaterno 's question: I think that things like bias and IA should still be specified as "tracer" systematics. We can write code that parses these and sets up the required dictionaries, etc. Baryons are a global systematic. This is the organization that was used in firecrown v0, and I think it made sense to the non-expert user.

**Comment by tilmantroester** [comment]
@damonge The `pyccl.CosmologyCalculator(cosmo, cosmo_args)` is pseudo code for filling out the relevant arguments, no need to change the API I think.
The reason for putting everything into the `pyccl.Cosmology` object is that this gets passed around everywhere. If there were a separate container for P(k)s in `TwoPoint` then that would need to be exposed so that other parts of the pipeline can access it.

@marcpaterno The specific 3x2pt systematics discussed here only concern sources and the `p_of_k_a` argument to `pyccl.angular_cl`. So that'd be limited to `GaussFamily`, I guess. Depending on the needs from SNe and clusters, a framework for more general systematics than those that apply to `Source` might be useful however.

@jablazek I reckon you could package bias and IA as source systematics but then you'd need to do extra book-keeping to check that the same power spectra are only computed once. But maybe I'm misunderstanding how these PT models are being setup. Which calculations are sources-specific and which apply to all sources of one kind (e.g. weak lensing)? An example with multiple tomographic bins might be helpful to understand this better.

**Comment by jablazek** [comment]
@tilmantroester : I agree - this is a question of bookkeeping. The setup for the PT models in CCL currently requires the PTCalculator object. But we could add a layer which initializes that based on the tracer properties being used (which are needed anyway to carry the IA/bias values). My main point is that that bias and IA are best conceptualized as tracer systematics, and we don't want to specify them with a different API depending on linear, nonlinear, emulation, etc.
